src/coreclr/tools/aot/ILCompiler/ILLinkDescriptorRootProvider.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+using AssemblyName = System.Reflection.AssemblyName;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// Compilation root provider that provides roots based on the ILLink descriptor file format.
+    /// Only supports a subset of the ILLink Descriptor file format.
+    /// </summary>
+    /// <remarks>https://github.com/mono/linker/blob/main/docs/data-formats.md</remarks>
+    internal class ILLinkDescriptorRootProvider : ICompilationRootProvider
```

Comment:
We have a base parser for the illink formats here: src\coreclr\tools\aot\ILCompiler.Compiler\Compiler\ProcessLinkerXmlBase.cs - this should ideally just derive from that class.

Reply:
@MichalStrehovsky this class is very simplistic and do not account for all possible options in the ILLink.Descriptor.xml
Usages which this parser does not support
- `<assembly preserve="all">`
- `<type fullName="XXX"/>` Does not include all methods in the processing.
- `<type preserve="nothing" />` or `<type preserve="all" />` too not supported.
Question following: 
- Should I try to account for these cases? 
- or I should create simple processing what's possible with current class and try to improve `ProcessLinkerXmlBase`?
- `ProcessLinkerXmlBase` has constructor parameter `_owningModule` which usage is not clear for me. Is this should be entry point for the application?

Reply:
We can add whatever virtual methods are needed into ProcessLinkerXmlBase. ILLink has a [similar class](https://github.com/mono/linker/blob/main/src/linker/Linker.Steps/ProcessLinkerXmlBase.cs) that this one is inspired by (we're just not using XPath because WTF). That class is also used as a base for everything.

`_owningModule` is used for the "embedded as a resource in the assembly" case. When the XML is embedded the `<assembly>` element is not mandatory and `<type>` can exist without an owning `<assembly>`. `_owningModule` would be null if the XML is not embedded.